From: Wei Liu <wei.liu2@citrix.com>
To: <xen-devel@lists.xen.org>, <netdev@vger.kernel.org>
CC: <ian.campbell@citrix.com>, <zoltan.kiss@citrix.com>,
	<david.vrabel@citrix.com>, Wei Liu <wei.liu2@citrix.com>
Subject: [PATCH net v4 1/3] xen-netback: move NAPI add/remove calls

Originally netif_napi_add was in xenvif_init_queue and netif_napi_del
was in xenvif_deinit_queue, while kthreads were handled in
xenvif_connect and xenvif_disconnect. Move netif_napi_add and
netif_napi_del to xenvif_connect and xenvif_disconnect so that they
reside together with kthread operations.

Signed-off-by: Wei Liu <wei.liu2@citrix.com>
Cc: Ian Campbell <ian.campbell@citrix.com>
Cc: Zoltan Kiss <zoltan.kiss@citrix.com>
---
diff --git a/drivers/net/xen-netback/interface.c b/drivers/net/xen-netback/interface.c
index 0c73a8b..4a34d5e 100644
--- a/drivers/net/xen-netback/interface.c
+++ b/drivers/net/xen-netback/interface.c
@@ -404,8 +404,6 @@ struct xenvif *xenvif_alloc(struct device *parent, domid_t domid,
 	memset(dev->dev_addr, 0xFF, ETH_ALEN);
 	dev->dev_addr[0] &= ~0x01;
 
-	netif_napi_add(dev, &vif->napi, xenvif_poll, XENVIF_NAPI_WEIGHT);
-
 	netif_carrier_off(dev);
 
 	err = register_netdev(dev);
@@ -505,6 +503,8 @@ int xenvif_connect(struct xenvif *vif, unsigned long tx_ring_ref,
 	wake_up_process(vif->task);
 	wake_up_process(vif->dealloc_task);
 
+	netif_napi_add(vif->dev, &vif->napi, xenvif_poll, XENVIF_NAPI_WEIGHT);
+
 	return 0;
 
 err_rx_unbind:
@@ -537,6 +537,8 @@ void xenvif_disconnect(struct xenvif *vif)
 {
 	xenvif_carrier_off(vif);
 
+	netif_napi_del(&vif->napi);
+
 	if (vif->task) {
 		del_timer_sync(&vif->rx_stalled);
 		kthread_stop(vif->task);
@@ -596,8 +598,6 @@ void xenvif_free(struct xenvif *vif)
 
 	free_xenballooned_pages(MAX_PENDING_REQS, vif->mmap_pages);
 
-	netif_napi_del(&vif->napi);
-
 	unregister_netdev(vif->dev);
 
 	vfree(vif->grant_copy_op);
